src/rgw/rgw_lc.cc

```diff
@@ -0,0 +1,546 @@
+  rados::cls::lock::Lock l(lc_index_lock_name);
+  do {
+    utime_t now = ceph_clock_now(g_ceph_context);
+    pair<string, int > entry;//string = bucket_name:bucket_id ,int = LC_BUCKET_STATUS
+    if (max_lock_secs <= 0)
+      return -EAGAIN;
+      
+    utime_t time(max_lock_secs, 0);
+    l.set_duration(time);
+
+    int ret = l.lock_exclusive(&store->lc_pool_ctx, obj_names[index]);
+    if (ret == -EBUSY) { /* already locked by another lc processor */
+      dout(0) << "RGWLC::process() failed to acquire lock on, sleep 5, try again" << obj_names[index] << dendl;
+      sleep(10);
+      continue;
+    }
+    if (ret < 0)
+      return 0;
+
+    string marker;
+    cls_rgw_lc_obj_head head;
+    ret = cls_rgw_lc_get_head(store->lc_pool_ctx, obj_names[index], head);
+    if (ret < 0) {
+      dout(0) << "RGWLC::process() failed to get obj head " << obj_names[index] << ret << dendl;
+      goto exit;
+    }
+
+    if(!if_already_run_today(head.start_date)) {
+      head.start_date = now;
+      head.marker.clear();
+      ret = bucket_lc_prepare(index);
+      if (ret < 0) {
+      dout(0) << "RGWLC::process() failed to update lc object " << obj_names[index] << ret << dendl;
+      goto exit;
+      }
+    }
+
+    ret = cls_rgw_lc_get_next_entry(store->lc_pool_ctx, obj_names[index], head.marker, entry);
+    if (ret < 0) {
+      dout(0) << "RGWLC::process() failed to get obj entry " << obj_names[index] << dendl;
+      goto exit;
+    }
+
+    if (entry.first.empty())
+      goto exit;
+
+    entry.second = lc_processing;
+    ret = cls_rgw_lc_set_entry(store->lc_pool_ctx, obj_names[index],  entry);
+    if (ret < 0) {
+      dout(0) << "RGWLC::process() failed to set obj entry " << obj_names[index] << entry.first << entry.second << dendl;
+      goto exit;
+    }
+
+    head.marker = entry.first;
+    ret = cls_rgw_lc_put_head(store->lc_pool_ctx, obj_names[index],  head);
+    if (ret < 0) {
+      dout(0) << "RGWLC::process() failed to put head " << obj_names[index] << dendl;
+      goto exit;
+    }
+    l.unlock(&store->lc_pool_ctx, obj_names[index]);
+    ret = bucket_lc_process(entry.first);
+    ret = bucket_lc_post(index, max_lock_secs, head, entry, ret);   
+    return 0;
+exit:    
+    l.unlock(&store->lc_pool_ctx, obj_names[index]);
+    return 0;
+    
+  }while(1);
+      
+}
+
+void RGWLC::start_processor()
+{
+  worker = new LCWorker(cct, this);
+  worker->create("lc_thread");
+}
+
+void RGWLC::stop_processor()
+{
+  if (worker) {
+    worker->stop();
+    worker->join();
+  }
+  delete worker;
+  worker = NULL;
+}
+
+void RGWLC::LCWorker::stop()
+{
+  Mutex::Locker l(lock);
+  cond.Signal();
+}
+
+bool RGWLC::going_down()
+{
+  return false;
+}
+
+bool RGWLC::LCWorker::should_work(utime_t& now)
+{
+  int start_hour;
+  int start_minite;
+  int end_hour;
+  int end_minite;
+  string worktime = cct->_conf->rgw_lifecycle_work_time;
+  sscanf(worktime.c_str(),"%d:%d-%d:%d",&start_hour, &start_minite, &end_hour, &end_minite);
```

Comment:
You could at least check the return value of sscanf(), and error out if it doesn't parse properly.